code/fallback/Hacl.IntTypes.Intrinsics.fst

```diff
@@ -58,10 +83,62 @@ let sub_borrow_st (t:inttype{t = U32 \/ t = U64}) =
     (let r = Seq.index (as_seq h1 r) 0 in
     v r - v c * pow2 (bits t) == v x - v y - v cin))
 
+#push-options "--z3rlimit 600"
 
 inline_for_extraction noextract
-val sub_borrow: #t:inttype{t = U32 \/ t = U64} -> sub_borrow_st t
-let sub_borrow #t cin x y r =
+val sub_borrow:
+  t:inttype{t = U32 \/ t = U64} ->
+  t':inttype{if t = U32 then t' = U64 else t' = U128} ->
+  shift_bits:UInt32.t{if t = U32 then shift_bits == 32ul else shift_bits == 64ul} ->
+  p:UInt32.t{if t = U32 then p == 64ul else p == 128ul} ->
+  sub_borrow_st t
+let sub_borrow t t' shift_bits p cin x y r =
+  let res = cast_to t t' x -. cast_to t t' y -. cast_to t t' cin in
+  assert (v res == ((v x - v y) % pow2 (v p) - v cin) % pow2 (v p));
+  assert (v res == (((v x - v y) @%. t') - v cin) @%. t');
+  Math.Lemmas.lemma_mod_add_distr (- v cin) (v x - v y) (pow2 (v p));
+  assert (v res == (v x - v y - v cin) @%. t');
+
+  assert (v res % pow2 (v shift_bits) = (v x - v y - v cin) % pow2 (v p) % pow2 (v shift_bits));
+  Math.Lemmas.pow2_modulo_modulo_lemma_1 (v x - v y - v cin) (v shift_bits) (v p);
+  assert (v res % pow2 (v shift_bits) = (v x - v y - v cin) % pow2 (v shift_bits));
+
+  let c = cast_from t' t (res >>. shift_bits) &. (to_uint t 1) in
+  assert_norm (pow2 1 = 2);
+  mod_mask_lemma (cast_from t' t (res >>. shift_bits)) 1ul;
+  assert (v ((mk_int #t #SEC 1 <<. 1ul) -! mk_int 1) == 1);
+  assert (v c = v res / pow2 (v shift_bits) % pow2 1);
+
+  r.(0ul) <- cast_from t' t res;
+  assert (v c = (if 0 <= v x - v y - v cin then 0 else 1));
+  c
+
+val sub_borrow_u32: sub_borrow_st U32
+let sub_borrow_u32 = sub_borrow U32 U64 32ul 64ul
+
+
+(* Fallback versions of add_carry_u64 and sub_borrow_u64 for platforms which
+   don't support uint128.
+   The names Hacl.IntTypes.Intrinsics.add_carry_u64 and sub_borrow_u64 must not
+   be changed because they are hardcoded in KreMLin for extracting wasm code
+   which uses these intrinsics. *)
+inline_for_extraction noextract
```

Comment:
I thought there *might* have been a use-case for an add-carry for 32-bit platforms without native uint64 but this seems unused here... so the `fallback` version does not rely on a wider type while the u64 version from this file does?

Reply:
I'm not sure I understand, are you saying that there is a requirement for having `add_carry_u32` and `sub_borrow_u32` implementations that don't rely on uint64? If so, I can keep both versions, similar to what I've done for `Hacl.IntTypes.Intrinsics_128.add_carry_u64`, which uses uint128, and `Hacl.IntTypes.Intrinsics.add_carry_u64`, which is just the current fallback version.

Reply:
yes, but maybe I'm thinking about a far-fetched use-case (a 32-bit embedded device or something) and we can re-instate those functions later on if the need arises